internal/cluster/connection_manager.go

```diff
@@ -0,0 +1,620 @@
+				if conn := m.connMap.GetConnectionForAddr(addrStr); conn == nil {
+					m.logger.Infof("found a broken connection to: %s, trying to fix it.", addrStr)
+					if addr, err := ParseAddress(addrStr); err != nil {
+						m.logger.Warnf("cannot parse address: %s", addrStr)
+					} else if err := m.connectAddr(addr); err != nil {
+						m.logger.Debug(func() string {
+							return fmt.Sprintf("cannot fix connection to %s: %s", addr, err.Error())
+						})
+					}
+				}
+			}
+			if m.connMap.Len() == 0 {
+				// if there are no connections, try to connect to seeds
+				if err := m.connectCluster(); err != nil {
+					m.logger.Errorf("while trying to fix cluster connection: %w", err)
+				}
+			}
+		}
+	}
+}
+
+type connectionMap struct {
+	connectionsMu *sync.RWMutex
+	// addrToConn maps connection address to connection
+	addrToConn map[string]*Connection
+	// connToAddr maps connection ID to address
+	connToAddr map[int64]*pubcluster.AddressImpl
+}
+
+func newConnectionMap() *connectionMap {
+	return &connectionMap{
+		connectionsMu: &sync.RWMutex{},
+		addrToConn:    map[string]*Connection{},
+		connToAddr:    map[int64]*pubcluster.AddressImpl{},
+	}
+}
+
+func (m *connectionMap) AddConnection(conn *Connection, addr *pubcluster.AddressImpl) {
+	m.connectionsMu.Lock()
+	defer m.connectionsMu.Unlock()
+	m.addrToConn[addr.String()] = conn
+	m.connToAddr[conn.connectionID] = addr
+}
+
+// RemoveConnection removes a connection and returns true if there are no more addrToConn left.
+func (m *connectionMap) RemoveConnection(removedConn *Connection) int {
+	m.connectionsMu.Lock()
+	defer m.connectionsMu.Unlock()
+	var remaining int
+	for addr, conn := range m.addrToConn {
+		if conn.connectionID == removedConn.connectionID {
+			delete(m.addrToConn, addr)
+			delete(m.connToAddr, conn.connectionID)
+			remaining = len(m.addrToConn)
+			break
+		}
+	}
+	return remaining
+}
+
+func (m *connectionMap) CloseAll() {
+	m.connectionsMu.RLock()
+	defer m.connectionsMu.RUnlock()
+	for _, conn := range m.addrToConn {
+		conn.close(nil)
+	}
+}
+
+func (m *connectionMap) GetConnectionForAddr(addr string) *Connection {
+	m.connectionsMu.RLock()
+	defer m.connectionsMu.RUnlock()
+	if conn, ok := m.addrToConn[addr]; ok {
+		return conn
+	}
+	return nil
+}
+
+func (m *connectionMap) GetAddrForConnectionID(connID int64) *pubcluster.AddressImpl {
+	m.connectionsMu.RLock()
+	defer m.connectionsMu.RUnlock()
+	if addr, ok := m.connToAddr[connID]; ok {
+		return addr
+	}
+	return nil
+}
+
+func (m *connectionMap) GetRandomAddr() *pubcluster.AddressImpl {
+	m.connectionsMu.RLock()
+	defer m.connectionsMu.RUnlock()
+	for _, addr := range m.connToAddr {
+		// get the first address
+		return addr
+	}
+	return nil
+}
+
+func (m *connectionMap) RandomConn() *Connection {
+	m.connectionsMu.RLock()
+	defer m.connectionsMu.RUnlock()
+	for _, conn := range m.addrToConn {
+		// Go randomizes maps, this is random enough for now.
```

Comment:
Maps in Go are based on hash tables, so the order on a stable map (i.e. when the are no mutations happening) will be the same among iterations. Hence, on for a stable cluster no randomization will take place. Not sure if that's what this comment implies, but other clients implement load balancers in order to "randomize" the selected connection.

See https://github.com/golang/go/blob/master/src/runtime/map.go#L23 for more details on how maps are implemented internally

Reply:
> the order on a stable map (i.e. when the are no mutations happening) will be the same among iterations. 
That's not correct. See: https://stackoverflow.com/questions/55925822/in-golang-why-are-iterations-over-maps-random

Some sample code:
```go
var m = map[string]int {
	"foo": 1,
	"bar": 2,
	"qoox": 3,
}

func random(m map[string]int) string {
	for k := range m {
		return k
	}
	return ""
}

func main() {
	for i := 0; i < 100; i++ {
		fmt.Print(random(m), " ")
	}
}
```
The output is not deterministic even though there is no mutation on the map.

The relevant code in the client is there until the load balancer is properly implemented.